P(X < oo) and P(X > oo) for True, False probability

gxyd opened the pull request:
fixes #10052

Commit:
P(X < oo) and P(X > oo) for True probability

fixes #10052

P(1) or similar(nonsensical) probabilities handled

given(x) for non-RadomSymbol `x` raise ValueError instead returning None

test-case added and error message corrected

jksuom commented:
This looks good.

There are other (nonsensical) cases where ``pspace(condition)`` returns ``None`` and ``AttributeError`` results, for example ``P(1)``. Perhaps these should also be caught and something more informative be returned (maybe ``nan``).
```

asmeurer commented on sympy/stats/rv.py:
```diff
@@ -608,6 +609,13 @@ def probability(condition, given_condition=None, numsamples=None,
     5/12
     """
 
+    if given_condition is S.false:
```
Better to use `==` than `is` here.

gxyd replied:
Your concern seems right to me.

gxyd commented:
> Perhaps these should also be caught and something more informative be returned (maybe nan).

Yes, i have added `ValueError`, but since `condition` is expected to be a `Relational`, so i think returning `ValueError` looks better to me instead of `nan`.

jksuom commented:
It is a good idea to catch any ``condition`` that is not a ``Relational``.
In addition, I think it would be advisable to make sure that ``pspace(condition)`` does not return ``None``. That can happen even when ``condition`` is a ``Relational`` if it does not contain random variables. For example, ``P(x < 1)`` for an ordinary symbol ``x``.

gxyd commented:
> It is a good idea to catch any condition that is not a Relational.

It can even take non-relationals like: `And(X + Y < 3, Y < 2) `. That's why
the current tests fail. I think i should use some, `.has(Relational)`
property  for that may be.

On Wed, Oct 28, 2015 at 12:41 PM, Kalevi Suominen <notifications@github.com>
wrote:

> It is a good idea to catch any condition that is not a Relational.
> In addition, I think it would be advisable to make sure that
> pspace(condition) does not return None. That can happen even when
> condition is a Relational if it does not contain random variables. For
> example, P(x < 1) for an ordinary symbol x.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/10053#issuecomment-151749233>.
>

gxyd commented:
Ex. P(And(X + Y < 2, X > 1)) where `X` and `Y` are both `RandomSymbol`.

On Wed, Oct 28, 2015 at 12:46 PM, Gaurav Dhingra <axyd0000@gmail.com> wrote:

> > It is a good idea to catch any condition that is not a Relational.
>
> It can even take non-relationals like: `And(X + Y < 3, Y < 2) `. That's
> why the current tests fail. I think i should use some, `.has(Relational)`
> property  for that may be.
>
> On Wed, Oct 28, 2015 at 12:41 PM, Kalevi Suominen <
> notifications@github.com> wrote:
>
>> It is a good idea to catch any condition that is not a Relational.
>> In addition, I think it would be advisable to make sure that
>> pspace(condition) does not return None. That can happen even when
>> condition is a Relational if it does not contain random variables. For
>> example, P(x < 1) for an ordinary symbol x.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/sympy/sympy/pull/10053#issuecomment-151749233>.
>>
>
>

jksuom commented:
Yes, there are ``BooleanFunction``s like ``And`` that should be accepted. Perhaps everything that derives from ``Boolean`` should be allowed. That would simplify the test.

gxyd commented:
I too think it is better for `given(condition)` to raise an Error instead of returning `None`. It is now ready for review.

gxyd commented:
I have addressed all the comments, though I `rebased` the PR, which led to the last two comments you mentioned to be removed(i have addressed them as well), since they were commented on PR-commits. Sorry for that.

jksuom commented:
No problem. The comments were only for temporary use.

asmeurer commented:
PyPy seems to be having some problems. It looks unrelated to this PR.

jksuom commented:
Just one job. I can restart it.

jksuom commented:
I think this is ready. I'll merge in 12 hours.

gxyd commented:
Thanks @jksuom @asmeurer
